File: src/connectivity/network/testing/netemul/runner/model/launch_service.cc
// Copyright 2018 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "launch_service.h"

#include "src/lib/pkg_url/fuchsia_pkg_url.h"

namespace netemul {
namespace config {

LaunchService::LaunchService(std::string name) : name_(std::move(name)) {}

bool LaunchService::ParseFromJSON(const rapidjson::Value& value,
                                  json::JSONParser* parser) {
  return launch_.ParseFromJSON(value, parser);
}

const std::string& LaunchService::name() const { return name_; }

const LaunchApp& LaunchService::launch() const { return launch_; }

}  // namespace config
}  // namespace netemul

File: src/experiences/session_shells/ermine/shell/lib/src/services/launch_service.dart
// Copyright 2021 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'package:fidl_fuchsia_element/fidl_async.dart';
import 'package:fuchsia_services/services.dart';

typedef ControllerClosedCallback = void Function(String id);

/// Namespace for annotations set by ermine.
const String ermineNamespace = 'ermine';

/// Namespace for annotations set by [ElementManager].
const String elementManagerNamespace = 'element_manager';

/// Defines a service to launch applications given their [url] using the
/// [ElementManager.proposeElement] API.
class LaunchService {
  late final ControllerClosedCallback onControllerClosed;

  Future<ControllerProxy> launch(String title, String url,
      {String? alternateServiceName}) async {
    final elementController = ControllerProxy();
    final proxy = ManagerProxy();

    final incoming = Incoming.fromSvcPath();
    if (alternateServiceName != null) {
      incoming.connectToServiceByNameWithChannel(
          alternateServiceName, proxy.ctrl.request().passChannel());
    } else {
      incoming.connectToService(proxy);
    }

    final id = '${DateTime.now().millisecondsSinceEpoch}';
    final annotations = [
      Annotation(
        key: AnnotationKey(namespace: ermineNamespace, value: 'id'),
        value: AnnotationValue.withText(id),
      ),
      Annotation(
        key: AnnotationKey(namespace: elementManagerNamespace, value: 'url'),
        value: AnnotationValue.withText(url),
      ),
      Annotation(
        key: AnnotationKey(namespace: ermineNamespace, value: 'name'),
        value: AnnotationValue.withText(title),
      ),
    ];

    final spec = Spec(componentUrl: url, annotations: annotations);
    await proxy.proposeElement(spec, elementController.ctrl.request());

    proxy.ctrl.close();
    await incoming.close();

    // ignore: unawaited_futures
    elementController.ctrl.whenClosed.then((_) => onControllerClosed(id));

    return elementController;
  }
}
